GetExtendedNPEMessage skips hidden frames and recognizes MH newInstance

JasonFengJ9 opened the pull request:
The `getStackTraceElementIterator()` invoked by `GetExtendedNPEMessage()` skips hidden frames;
`JDK18` `NPE` extended message generation recognizes `JEP416` `MH` methods for `java.lang.reflect.Constructor.newInstance()`;
Updated `test_getCallerClass_Helper_Reflection_fromBootExtWithAnnotation()` to adopt `MH` frames;
Removed `-Djdk.reflect.useDirectMethodHandle=false` to re-enable `JEP416`.

Note: currently OpenJ9 stack frame iterator functions check `J9AccMethodFrameIteratorSkip` individually, ideally a unified solution can be developed to hide/show those hidden frames. Created https://github.com/eclipse-openj9/openj9/issues/13951.

Signed-off-by: Jason Feng <fengj@ca.ibm.com>

gacholio commented on runtime/j9vm/javanextvmi.c:
```diff
@@ -70,14 +70,21 @@ typedef struct GetStackTraceElementUserData {
+		&& J9_ARE_ALL_BITS_SET(romMethod->modifiers, J9AccMethodFrameIteratorSkip)
+	) {
+		/* Skip methods with java.lang.invoke.FrameIteratorSkip / jdk.internal.vm.annotation.Hidden / java.lang.invoke.LambdaForm$Hidden annotation */
+		return J9_STACKWALK_KEEP_ITERATING;
```
Please avoid multiple returns.

JasonFengJ9 replied:
Fixed.

gacholio commented:
The UTF name comparisons are a truly awful way of implementing this. Could the methods not be put in the VM constant pool for direct comparison?

gacholio commented:
jenkins test sanity,extended win jdk18

pshipton commented:
jenkins test sanity,extended win jdknext

pshipton commented:
jenkins compile zlinux,plinux,alinux64,xlinux,aix,osx jdknext

pshipton commented:
Actually, since we're enabling JEP 416 here, I'm going to run sanity for all platforms.

jenkins test sanity zlinux,plinux,alinux64,xlinux,aix,osx jdknext

gacholio commented:
jenkins compile win jdk8

JasonFengJ9 commented:
> The UTF name comparisons are a truly awful way of implementing this. Could the methods not be put in the VM constant pool for direct comparison?

Looking into it.

pshipton commented:
There are some extended.functional failures, but probably not related to this PR. I can run it outside of this PR to compare.

https://openj9-jenkins.osuosl.org/view/Test/job/Grinder/406/

pshipton commented:
The AIX failure is an unrelated known issue.

pshipton commented:
The same extended.functional failures occur without this PR.

The xlinux failure is crash in the JIT library with vmState=0x00000000. Hard to say if we've seen the crash before, there isn't much to go on. I don't find a report of a similar crash in the same test. Seems unlikely to be related to this change.

JasonFengJ9 commented:
> The UTF name comparisons are a truly awful way of implementing this. Could the methods not be put in the VM constant pool for direct comparison?

`iterateStackTrace` used by NPE extended message computing returns `J9ROMMethod`. The method such as `J9VMJDKINTERNALREFLECTDIRECTCONSTRUCTORHANDLEACCESSOR_INVOKEIMPL_METHOD` generated from `vmconstantpool.xml` is `J9Method` w/ its own `J9ROMMethod` which is different from the ROM method returned from `iterateStackTrace` callback even though they refer to same `class`/`method`/`sig` as per `UTF` name comparisons.
Note: `J9Method` can be used for comparison indeed but there is no straightforward way to get one via `iterateStackTrace` .
Any suggestions?

gacholio commented:
The VM constant pool method must be the same method as you get in the stack traces or there's a much bigger problem here. Please investigate why there is a difference.

gacholio commented:
The method change can wait for a future PR, assuming it can be done.

gacholio commented:
One possible reason for the method changing is HCR. The VMCP should always have the most up to date method, but stack traces may well be running an older version. To get around this, we would need to loop over all of the replaced methods, which makes this solution a lot less appealing.

JasonFengJ9 commented:
**PR test failures:**

* cmdLineTest_J9test_common_0_FAILED [1]
  * This is a known issue https://github.com/eclipse-openj9/openj9/issues/13577
* cmdLineTester_callsitedbgddrext_openj9_0_FAILED [2]
  * Reproducible w/o this PR, created https://github.com/eclipse-openj9/openj9/issues/13964 
* cmdLineTester_modularityddrtests_1_FAILED [3]
  * Reproducible w/o this PR, created https://github.com/eclipse-openj9/openj9/issues/13965 

None of PR test failures are caused by this PR.

[1] https://openj9-jenkins.osuosl.org/job/Test_openjdknext_j9_sanity.functional_ppc64_aix_Personal/8/tapResults/
[2] https://openj9-jenkins.osuosl.org/job/Test_openjdknext_j9_sanity.functional_x86-64_linux_Personal/10/tapResults/
[3] https://openj9-jenkins.osuosl.org/job/Test_openjdknext_j9_extended.functional_x86-64_windows_Personal/1/tapResults/

JasonFengJ9 commented:
> One possible reason for the method changing is HCR. The VMCP should always have the most up to date method, but stack traces may well be running an older version. To get around this, we would need to loop over all of the replaced methods, which makes this solution a lot less appealing.

For NPE message computing, the `romMethod` retrieved from `iterateStackTrace` is the method invoking another method which had NPE, i.e., the incoming ROM method is the caller, not the method having NPE.
https://github.com/eclipse-openj9/openj9/blob/fc2d3d10d53747b125d30f35b2ba9a4be1adeae0/runtime/vm/extendedMessageNPE.cpp#L786-L787
The method caused NPE is determined via 
https://github.com/eclipse-openj9/openj9/blob/fc2d3d10d53747b125d30f35b2ba9a4be1adeae0/runtime/vm/extendedMessageNPE.cpp#L527-L528
which has only `ROM` method `name`/`sig`.
To compare w/ VMCP methods, we need to loop over rom methods which is not desirable either.

> iterateStackTrace used by NPE extended message computing returns J9ROMMethod. The method such as J9VMJDKINTERNALREFLECTDIRECTCONSTRUCTORHANDLEACCESSOR_INVOKEIMPL_METHOD generated from vmconstantpool.xml is J9Method w/ its own J9ROMMethod which is different from the ROM method returned from iterateStackTrace callback even though they refer to same class/method/sig as per UTF name comparisons.

**Correction**: it compared incoming ROM method w/ VMCP method, and `name`/`sig` retrieved from `J9ROMMethodRef` w/ VMCP method `name`/`sig` which is invalid.
For experimental verification purpose, added the caller `class`/`method` into `vmconstantpool.xml`,  modified `exceptiondescribe.c:iterateStackTrace()` to return the RAM method, VMCP RAM/ROM methods match the methods returned from `iterateStackTrace`.
However comparing the caller method doesn't work. For example the caller of `java.lang.NullPointerException.<init>()` is an application method like `a.se.npe.TestNPE.testAll(TestNPE.java:50)`.
